modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFileEntryMVCActionCommand.java

```diff
@@ -116,9 +138,166 @@ private void _copyFileEntry(
 		}
 	}
 
+	private ServiceContext _createServiceContext(
+			ActionRequest actionRequest, long fileEntryId, Group group)
+		throws PortalException {
+
+		ServiceContext serviceContext = ServiceContextFactory.getInstance(
+			DLFileEntry.class.getName(), actionRequest);
+
+		try {
+			AssetEntry assetEntry = _assetEntryService.getEntry(
+				DLFileEntry.class.getName(), fileEntryId);
+
+			List<Long> groupIds = _getGroupIds(
+				ListUtil.fromArray(
+					_portal.getCurrentAndAncestorSiteGroupIds(
+						group.getGroupId())),
+				group);
+
+			serviceContext.setAssetCategoryIds(
+				_getAssetCategoryIds(assetEntry, group, groupIds));
+
+			serviceContext.setAssetTagNames(
+				_getAssetTagNames(assetEntry, groupIds));
+		}
+		catch (PortalException portalException) {
+			if (!(portalException instanceof NoSuchEntryException)) {
+				throw portalException;
+			}
+		}
+
+		return serviceContext;
+	}
+
+	private long[] _getAssetCategoryIds(
+			AssetEntry assetEntry, Group group, List<Long> groupIds)
+		throws PortalException {
+
+		Set<Long> categoryIds = new HashSet<>();
+
+		for (long categoryId : assetEntry.getCategoryIds()) {
+			if (_isCategoryIdAllowed(
+					categoryId, groupIds, group,
+					_depotEntryLocalService.fetchGroupDepotEntry(
+						group.getGroupId()))) {
+
+				categoryIds.add(categoryId);
+			}
+		}
+
+		return ArrayUtil.toLongArray(categoryIds);
+	}
+
+	private String[] _getAssetTagNames(
+		AssetEntry assetEntry, List<Long> groupIds) {
+
+		Set<String> tagNames = new HashSet<>();
+
+		for (String tagName : assetEntry.getTagNames()) {
+			if (_isTagNameAllowed(groupIds, tagName)) {
+				tagNames.add(tagName);
+			}
+		}
+
+		return ArrayUtil.toStringArray(tagNames);
+	}
+
+	private List<Long> _getGroupIds(
```

Comment:
Cannot we use `com.liferay.depot.internal.group.provider.SiteConnectedGroupGroupProviderImpl#getCurrentAndAncestorSiteAndDepotGroupIds(long)` here?